packages/grid/x-grid-modules/src/hooks/features/useColumnReorder.tsx

```diff
@@ -0,0 +1,65 @@
+import * as React from 'react';
+import { ColDef } from '../../models/colDef';
+import { useLogger } from '../utils';
+import { ApiRef } from '../../models';
+
+export const useColumnReorder = (
+  apiRef: ApiRef
+) => {
+  const logger = useLogger('useColumnReorder');
+
+  const dragCol = React.useRef<ColDef | null>();
+  const dragColNode = React.useRef<HTMLElement | null>();
+
+  const handleDragStart = React.useCallback(
+    (col: ColDef, htmlEl: any): void => {
+      logger.debug(`Start dragging col ${col.field}`);
+      dragCol.current = col;
+      dragColNode.current = htmlEl;
+      dragColNode.current?.addEventListener('dragend', handleDragEnd);
+      dragColNode.current?.classList.add('dragging');
+      setTimeout(() => {
+        dragColNode.current?.classList.remove('dragging');
+      }, 0);
+    },
+    [apiRef, logger]
+  );
+
+  const handleDragEnter = React.useCallback(
+    (col: ColDef): void => {
+      logger.debug(`Enter dragging col ${col.field}`);
+
+      if (col.field !== dragCol.current?.field) {
+        const targetColIndex = apiRef.current.getColumnIndex(col.field);
+        const dragColIndex = apiRef.current.getColumnIndex(dragCol.current!.field);
+        const columnsSnapshot = apiRef.current.getAllColumns();
+
+        apiRef.current.scrollToIndexes({
+          colIndex: targetColIndex,
+          rowIndex: 0
+        });
+
+        columnsSnapshot.splice(targetColIndex, 0, columnsSnapshot.splice(dragColIndex, 1)[0]);
+
```

Comment:
Same as my comment on `onColumnDragStart`, here we can publish an event for columns reordering and create optional handler that the user can hook into, maybe `onColumnsReordering` or `onColumnsOrderChange`? 🤔

Reply:
If you want people to be able to change the reordering logic then yes. The only thing this does is to remove an element from the array and put it on another place in the same array and then shift all the other elements right by 1.